using global with name 'array' causes c compiler error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.7 18eee34.2e0fa12
OS: windows, Microsoft Windows 11 Pro v22631 64-bit
Processor: 24 cpus, 64bit, little endian, 

getwd: D:\vbug
vexe: C:\v\v.exe
vexe mtime: 2024-09-28 12:22:23

vroot: OK, value: C:\v
VMODULES: OK, value: C:\Users\info\.vmodules
VTMP: OK, value: C:\Users\info\AppData\Local\Temp\v_0

Git version: git version 2.42.0.windows.2
Git vroot status: weekly.2024.39-36-g2e0fa122-dirty
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 b425ac82

```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg src/main.v`
```v
module main

__global (
	array [16]int 
)

fn main() {}

```

**What did you expect to see?**

no c error


**What did you see instead?**
```
v -enable-globals run .
================== C compilation error (from tcc): ==============
cc: C:/Users/info/AppData/Local/Temp/v_0/vbug.01J8WBAS08MR0B9N5SKKGS4PY4.tmp.c:526: warning: WINVER redefined
cc: C:/Users/info/AppData/Local/Temp/v_0/vbug.01J8WBAS08MR0B9N5SKKGS4PY4.tmp.c:2166: error: redefinition of 'array'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -10,6 +10,11 @@ import v.util
 import v.pref
 import sync.stdatomic
 
+// V type names that cannot be used as global var name
+pub const global_reserved_type_names = ['byte', 'bool', 'char', 'i8', 'i16', 'int', 'i64', 'u8',
+	'u16', 'u32', 'u64', 'f32', 'f64', 'map', 'string', 'rune', 'usize', 'isize', 'voidptr', 'thread',
+	'array']
+
 pub type TypeDecl = AliasTypeDecl | FnTypeDecl | SumTypeDecl
 
 // pub const int_type_name = $if amd64 || arm64 {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -33,6 +33,9 @@ pub const array_builtin_methods = ['filter', 'clone', 'repeat', 'reverse', 'map'
 	'pop', 'delete']
 pub const array_builtin_methods_chk = token.new_keywords_matcher_from_array_trie(array_builtin_methods)
 // TODO: remove `byte` from this list when it is no longer supported
+pub const global_reserved_type_names = ['byte', 'bool', 'char', 'i8', 'i16', 'int', 'i64', 'u8',
+	'u16', 'u32', 'u64', 'f32', 'f64', 'map', 'string', 'rune', 'usize', 'isize', 'voidptr', 'thread',
+	'array']
 pub const reserved_type_names = ['byte', 'bool', 'char', 'i8', 'i16', 'int', 'i64', 'u8', 'u16',
 	'u32', 'u64', 'f32', 'f64', 'map', 'string', 'rune', 'usize', 'isize', 'voidptr', 'thread']
 pub const reserved_type_names_chk = token.new_keywords_matcher_from_array_trie(reserved_type_names)
@@ -2302,6 +2305,11 @@ fn (mut c Checker) global_decl(mut node ast.GlobalDecl) {
 	}
 	for mut field in node.fields {
 		c.check_valid_snake_case(field.name, 'global name', field.pos)
+
+		if field.name in global_reserved_type_names {
+			c.error('invalid use of reserved type `${field.name}` as a global name', field.pos)
+		}
+
 		if field.name in c.global_names {
 			c.error('duplicate global `${field.name}`', field.pos)
 		}

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -33,9 +33,6 @@ pub const array_builtin_methods = ['filter', 'clone', 'repeat', 'reverse', 'map'
 	'pop', 'delete']
 pub const array_builtin_methods_chk = token.new_keywords_matcher_from_array_trie(array_builtin_methods)
 // TODO: remove `byte` from this list when it is no longer supported
-pub const global_reserved_type_names = ['byte', 'bool', 'char', 'i8', 'i16', 'int', 'i64', 'u8',
-	'u16', 'u32', 'u64', 'f32', 'f64', 'map', 'string', 'rune', 'usize', 'isize', 'voidptr', 'thread',
-	'array']
 pub const reserved_type_names = ['byte', 'bool', 'char', 'i8', 'i16', 'int', 'i64', 'u8', 'u16',
 	'u32', 'u64', 'f32', 'f64', 'map', 'string', 'rune', 'usize', 'isize', 'voidptr', 'thread']
 pub const reserved_type_names_chk = token.new_keywords_matcher_from_array_trie(reserved_type_names)
@@ -2306,7 +2303,7 @@ fn (mut c Checker) global_decl(mut node ast.GlobalDecl) {
 	for mut field in node.fields {
 		c.check_valid_snake_case(field.name, 'global name', field.pos)
 
-		if field.name in global_reserved_type_names {
+		if field.name in ast.global_reserved_type_names {
 			c.error('invalid use of reserved type `${field.name}` as a global name', field.pos)
 		}
 

--- a/vlib/v/parser/parser.v
+++ b/vlib/v/parser/parser.v
@@ -18,6 +18,9 @@ import strings
 const allowed_lock_prefix_ins = ['add', 'adc', 'and', 'btc', 'btr', 'bts', 'cmpxchg', 'cmpxchg8b',
 	'cmpxchg16b', 'dec', 'inc', 'neg', 'not', 'or', 'sbb', 'sub', 'xor', 'xadd', 'xchg']
 
+const reserved_type_names = ['byte', 'bool', 'char', 'i8', 'i16', 'int', 'i64', 'u8', 'u16', 'u32',
+	'u64', 'f32', 'f64', 'map', 'string', 'rune', 'usize', 'isize', 'voidptr', 'thread', 'array']
+
 @[minify]
 pub struct Parser {
 pub:
@@ -4059,6 +4062,11 @@ fn (mut p Parser) global_decl() ast.GlobalDecl {
 		}
 		pos := p.tok.pos()
 		name := p.check_name()
+
+		if name in reserved_type_names {
+			p.error_with_pos('invalid use of reserved type `${name}` as a global name',
+				pos)
+		}
 		has_expr := p.tok.kind == .assign
 		mut expr := ast.empty_expr
 		mut typ := ast.void_type

--- a/vlib/v/parser/parser.v
+++ b/vlib/v/parser/parser.v
@@ -4062,11 +4062,6 @@ fn (mut p Parser) global_decl() ast.GlobalDecl {
 		}
 		pos := p.tok.pos()
 		name := p.check_name()
-
-		if name in reserved_type_names {
-			p.error_with_pos('invalid use of reserved type `${name}` as a global name',
-				pos)
-		}
 		has_expr := p.tok.kind == .assign
 		mut expr := ast.empty_expr
 		mut typ := ast.void_type
@@ -4114,7 +4109,9 @@ fn (mut p Parser) global_decl() ast.GlobalDecl {
 			is_exported: is_exported
 		}
 		fields << field
-		p.table.global_scope.register(field)
+		if name !in reserved_type_names {
+			p.table.global_scope.register(field)
+		}
 		comments = []
 		if !is_block {
 			break

--- a/vlib/v/parser/parser.v
+++ b/vlib/v/parser/parser.v
@@ -18,9 +18,6 @@ import strings
 const allowed_lock_prefix_ins = ['add', 'adc', 'and', 'btc', 'btr', 'bts', 'cmpxchg', 'cmpxchg8b',
 	'cmpxchg16b', 'dec', 'inc', 'neg', 'not', 'or', 'sbb', 'sub', 'xor', 'xadd', 'xchg']
 
-const reserved_type_names = ['byte', 'bool', 'char', 'i8', 'i16', 'int', 'i64', 'u8', 'u16', 'u32',
-	'u64', 'f32', 'f64', 'map', 'string', 'rune', 'usize', 'isize', 'voidptr', 'thread', 'array']
-
 @[minify]
 pub struct Parser {
 pub:
@@ -4109,7 +4106,7 @@ fn (mut p Parser) global_decl() ast.GlobalDecl {
 			is_exported: is_exported
 		}
 		fields << field
-		if name !in reserved_type_names {
+		if name !in ast.global_reserved_type_names {
 			p.table.global_scope.register(field)
 		}
 		comments = []

--- a/vlib/v/parser/tests/global_reserved_name_err.out
+++ b/vlib/v/parser/tests/global_reserved_name_err.out
@@ -1,0 +1,7 @@
+vlib/v/parser/tests/global_reserved_name_err.vv:5:2: error: invalid use of reserved type `array` as a global name
+    3 | 
+    4 | __global (
+    5 |     array [16]int
+      |     ~~~~~
+    6 | )
+    7 |

--- a/vlib/v/parser/tests/global_reserved_name_err.vv
+++ b/vlib/v/parser/tests/global_reserved_name_err.vv
@@ -6,5 +6,4 @@ __global (
 )
 
 fn main() {
-	assert array.len == 16
 }

--- a/vlib/v/parser/tests/global_reserved_name_err.vv
+++ b/vlib/v/parser/tests/global_reserved_name_err.vv
@@ -1,0 +1,10 @@
+@[has_globals]
+module main
+
+__global (
+	array [16]int
+)
+
+fn main() {
+	assert array.len == 16
+}

